hydra-node/src/Hydra/Chain.hs

```diff
@@ -220,7 +248,7 @@ data ChainEvent tx
       , newChainState :: ChainStateType tx
       }
   | Rollback
-      { rolledBackChainState :: ChainStateType tx
+      { rolledBackChainState :: ChainStateHistory tx
```

Comment:
It's a bit weird that we provide the full history on the `Rollback` event. In a way, this is shifting the full history from the recursive type to shifting the whole history (as a list) here.

I think we should keep this a `ChainStateType` here.

In fact, the single usage site of `Rollback` event in our production code will (via the `ChainRolledBack` event) just use `setChainStateHistory` and that function only uses `currentState` of it!

Reply:
This will further complicate the `recoverChainStateHistory` function, but if you think about it, the information is already present in the persisted `StateChanged` events.

We originally only had a `Rollback ChainSlot` event here and maybe this is the most minimal we can get by again?

There is a trade-off with my suggestion: we would trade redundant information in the event store with duplicating the logic for "how to rollback". That is, walking into the past and finding the first `ChainStateType` with `<= rollbackSlot`